### PR TITLE
Improved error message for training regression on inf/NaN (MLDB-1605)

### PR DESCRIPTION
**Pull request by jeremybarnes**
Also includes some more functions we can use in the WHERE to fix it.

### CODE REVIEW

File: container_files/public_html/doc/builtin/sql/ValueExpression.md [hunk 1]
```diff
@@ -365,6 +365,10 @@ With `{arrays: 'encode'}` the output will be:
 - `mod(x, y)`: returns x modulo y.  The value of x and y must be an integer.
 - `abs(x)`: returns the absolute value of x.
 - `sqrt(x)`: returns the square root of x.  The value of x must be greater or equal to 0.
+- `isnan(x)`: return true if x is not-a-number in the floating point representation.
```
Comment (nicolaskruchten) [inline container_files/public_html/doc/builtin/sql/ValueExpression.md]: would be good to have underscores to match replace_*

**Comment by jeremybarnes** [comment]
I added replace_null to the PR also.

**Comment by nicolaskruchten** [comment]
Do we also need is_nad() for detecting not-a-date? Or can I pass a
timestamp into is_nan() and it will coerce and return the right thing?

On Saturday, 30 April 2016, Jeremy Barnes <notifications@github.com> wrote:

> I added replace_null to the PR also.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/mldbai/mldb/pull/343#issuecomment-215960052>
>


-- 
Nicolas Kruchten, Eng  |  Staff Software Engineer at Datacratic  |
http://datacratic.com/

**Comment by jeremybarnes** [comment]
It doesn't handle timestamps.  We could add an equivalent function, but maybe that should be another DOD.

**Comment by nicolaskruchten** [comment]
+1

File: plugins/classifier.cc [hunk 2]
```diff
@@ -528,6 +528,15 @@ run(const ProcedureRunConfig & run,
         if (!isfinite(weight))
             throw HttpReturnException(400, "classifier example weights must be finite");
 
+        if (runProcConf.mode == CM_REGRESSION
+            && !isfinite(label)) {
+            throw HttpReturnException
+                (400,
+                 "Regression labels must not be infinite or NaN.  Should you "
+                 "add a condition like `WHERE isfinite(label)` to your data, "
+                 "or preprocess your labels with `replace_if_not_finite(label, 0)`?");
```
Comment (guyd) [inline plugins/classifier.cc]: The name of the newly added function is `replace_not_finite`.

Comment (nicolaskruchten) [inline plugins/classifier.cc]: i'll hot-patch the error message.

Comment (nicolaskruchten) [inline plugins/classifier.cc]: done. nice catch @guyd

**Pull request merged**